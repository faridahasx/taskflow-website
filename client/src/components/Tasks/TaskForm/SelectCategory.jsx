// External imports
import PropTypes from "prop-types";
// MUI Components
import FormControl from "@mui/material/FormControl";
import Select from "@mui/material/Select";
import InputLabel from "@mui/material/InputLabel";
import MenuItem from "@mui/material/MenuItem";

const SelectCategory = (props) => {
  // Destructure props
  const { handleInputChange, category, categories } = props;

  return (
    <FormControl sx={{ width: "100%" }}>
      <InputLabel id="category">Category</InputLabel>
      <Select
        id="category"
        labelId="category"
        value={category || "All"}
        name="category"
        label="category"
        onChange={handleInputChange}
      >
        {categories.map((cat) => (
          <MenuItem value={cat.title} key={cat._id}>
            {cat.title}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};

SelectCategory.propTypes = {
  category: PropTypes.string.isRequired,
  categories: PropTypes.array.isRequired,
  handleInputChange: PropTypes.func.isRequired,
};

export default SelectCategory;
