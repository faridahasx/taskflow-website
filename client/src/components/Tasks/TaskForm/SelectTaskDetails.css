#select-task-details {
    justify-content: space-between;
    padding-top: 10px;
}

.select-task-detail {
    width: 33%;
    min-width: 33%;
}