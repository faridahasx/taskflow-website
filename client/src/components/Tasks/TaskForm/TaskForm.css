#task-form {
    height: calc(var(--vh) - 70px);
    justify-content: space-around;
}

#task-form .task-dialog-title {
    max-height: 40px;
    padding: 0 20px;
    background-color: rgb(234, 234, 234);
    font-size: 15px;
}

#task-form .task-dialog-title:focus-within {
    background-color: rgb(255, 255, 255);
}


#task-title-input:focus-within {
    background-color: white;
}

.description-input-loading {
    min-height: calc(var(--vh) - 300px);
}

@media all and (min-width: 900px) and (min-height: 500px) {
    .description-input-loading {
        min-height: calc(((var(--vh) / 100) * 88) - 300px);
    }

    #task-form {
        height: calc(((var(--vh) / 100) * 88) - 70px);
    }

}