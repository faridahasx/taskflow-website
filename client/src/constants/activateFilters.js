const activateFilters = [
  { buttonTxt: "All", value: "all" },
  { buttonTxt: "Current", value: "current" },
  { buttonTxt: "Overdue", value: "overdue" },
  { buttonTxt: "Upcoming", value: "upcoming" },
  { buttonTxt: "Completed", value: "completed" },
];

export default activateFilters;
