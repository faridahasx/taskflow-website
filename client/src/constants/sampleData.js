import { getDateWithNDaysOfDifference } from "../utils/dateUtils";

const categoriesSample = [
  {
    title: "All",
    tasks: 5,
    _id: "All",
  },
  {
    _id: "Tasks",
    tasks: 5,
    title: "Tasks",
  },
];

let tasksSample = [
  {
    _id: "1",
    title: "Task 1",
    category: "Tasks",
    description: "",
    startDate: getDateWithNDaysOfDifference(-1),
    finishDate: getDateWithNDaysOfDifference(1),
    completedAt: null,
    createdAt: getDateWithNDaysOfDifference(-4),
    updatedAt: getDateWithNDaysOfDifference(-4),
    __v: 0,
  },
  {
    _id: "2",
    title: "Task 2",
    category: "Tasks",
    description: "",
    startDate: getDateWithNDaysOfDifference(-2),
    finishDate: getDateWithNDaysOfDifference(-1),
    completedAt: null,
    createdAt: getDateWithNDaysOfDifference(-4),
    updatedAt: getDateWithNDaysOfDifference(-4),
    __v: 0,
  },
  {
    _id: "3",
    title: "Task 3",
    category: "Tasks",
    description: "",
    startDate: getDateWithNDaysOfDifference(2),
    finishDate: getDateWithNDaysOfDifference(3),
    completedAt: null,
    createdAt: getDateWithNDaysOfDifference(-4),
    updatedAt: getDateWithNDaysOfDifference(-4),
    __v: 0,
  },

  {
    _id: "4",
    title: "Task 4",
    category: "Tasks",
    description: "",
    startDate: getDateWithNDaysOfDifference(2),
    finishDate: getDateWithNDaysOfDifference(4),
    completedAt: getDateWithNDaysOfDifference(3),
    createdAt: getDateWithNDaysOfDifference(-4),
    updatedAt: getDateWithNDaysOfDifference(-4),
    __v: 0,
  },
  {
    _id: "5",
    title: "Login to continue",
    category: "Tasks",
    description: "",
    startDate: getDateWithNDaysOfDifference(-1),
    finishDate: getDateWithNDaysOfDifference(1),
    completedAt: null,
    createdAt: getDateWithNDaysOfDifference(-4),
    updatedAt: getDateWithNDaysOfDifference(-4),
    __v: 0,
  },
];

export { tasksSample, categoriesSample };
