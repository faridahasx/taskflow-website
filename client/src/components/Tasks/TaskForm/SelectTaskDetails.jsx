// External imports
import PropTypes from "prop-types";
import dayjs from "dayjs";
// Components
import DatePicker from "../../DatePicker";
import SelectCategory from "./SelectCategory";
// Styles
import "./SelectTaskDetails.css";

const SelectTaskDetails = (props) => {
  // Destructure props
  const {
    handleInputChange,
    handleSetStartDate,
    handleSetFinishDate,
    category,
    startDate,
    finishDate,
    categories,
  } = props;

  return (
    <div id="select-task-details" className="flex">
      <div className="select-task-detail">
        <SelectCategory
          handleInputChange={handleInputChange}
          category={category}
          categories={categories}
        />
      </div>
      <div className="select-task-detail">
        <DatePicker
          name="startDate"
          label="Start Date"
          date={startDate}
          setDate={handleSetStartDate}
          sx={{ width: "100%" }}
        />
      </div>
      <div className="select-task-detail">
        <DatePicker
          name="finishDate"
          label="Finish Date"
          date={finishDate}
          setDate={handleSetFinishDate}
          sx={{ width: "100%" }}
        />
      </div>
    </div>
  );
};

SelectTaskDetails.propTypes = {
  handleInputChange: PropTypes.func.isRequired,
  handleSetStartDate: PropTypes.func.isRequired,
  handleSetFinishDate: PropTypes.func.isRequired,
  categories: PropTypes.array.isRequired,
  category: PropTypes.string.isRequired,
  startDate: PropTypes.instanceOf(dayjs),
  finishDate: PropTypes.instanceOf(dayjs),
};

export default SelectTaskDetails;
