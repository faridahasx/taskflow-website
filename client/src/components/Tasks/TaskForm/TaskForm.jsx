// External imports
import { useState } from "react";
import PropTypes from "prop-types";
import dayjs from "dayjs";
// Custom hooks
import useBodyOverflowHidden from "../../../hooks/useBodyOverflowHidden";
// Components
import SubmitButton from "../../IconButtons/SubmitButton";
import ConfirmationDialog from "../../ConfirmationDialog/ConfirmationDialog";
import CircularLoading from "../../Loading/CircularLoading";
import Editor from "./TextEditor/Editor";
import TaskDialog from "../TaskDialog";
import SelectTaskDetails from "./SelectTaskDetails";
// Styles
import "./TaskForm.css";

const TaskForm = (props) => {
  // Destructure props
  const {
    handleCloseEditor,
    handleSubmit,
    handleInputChange,
    title,
    category,
    startDate,
    handleSetStartDate,
    finishDate,
    handleSetFinishDate,
    description,
    handleSetDescription,
    loading,
    saved,
    categories,
  } = props;
  // Custom hook to handle body overflow
  useBodyOverflowHidden();
  // Local state for cancel confirmation
  const [showCancelCloseDialog, setShowCancelCloseDialog] = useState(false);

  const handleShowCloseEditorOptions = () => {
    saved ? handleCloseEditor() : setShowCancelCloseDialog(true);
  };

  // Handle close button click
  const handleCloseClick = (e) => {
    e.preventDefault();
    // Check if changes are saved before closing
    handleShowCloseEditorOptions();
  };

  // Handle cancel close action
  const handleCancelClose = () => {
    setShowCancelCloseDialog(false);
  };

  return (
    <TaskDialog
      handleBackgroundClick={handleShowCloseEditorOptions}
      handleCloseClick={handleCloseClick}
      SecondButton={
        <SubmitButton
          loading={loading}
          buttonProps={{
            form: "task-form",
            title: "Save task",
            disabled: loading,
          }}
        />
      }
    >
      <form onSubmit={handleSubmit} id="task-form" className="column">
        <input
          className="task-dialog-title"
          maxLength={200}
          required
          autoFocus
          placeholder="Title..."
          onChange={handleInputChange}
          type="text"
          name="title"
          value={title}
        />

        {description === undefined ? (
          <div className="description-input-loading center">
            <CircularLoading />
          </div>
        ) : (
          <Editor
            description={description}
            handleSetDescription={handleSetDescription}
          />
        )}
        <SelectTaskDetails
          handleInputChange={handleInputChange}
          handleSetStartDate={handleSetStartDate}
          handleSetFinishDate={handleSetFinishDate}
          category={category}
          startDate={startDate}
          finishDate={finishDate}
          categories={categories}
        />
      </form>

      {showCancelCloseDialog && (
        <ConfirmationDialog
          handleCloseDialog={handleCancelClose}
          handleCancel={handleCancelClose}
          cancelButtonText="Cancel"
          handleConfirm={handleCloseEditor}
          confirmButtonText="Exit"
          loading={loading}
          heading="Your draft will be lost"
        />
      )}
    </TaskDialog>
  );
};

TaskForm.propTypes = {
  handleCloseEditor: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  handleInputChange: PropTypes.func.isRequired,
  handleSetStartDate: PropTypes.func.isRequired,
  handleSetFinishDate: PropTypes.func.isRequired,
  handleSetDescription: PropTypes.func.isRequired,
  category: PropTypes.string.isRequired,
  categories: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  description: PropTypes.string,
  title: PropTypes.string,
  startDate: PropTypes.instanceOf(dayjs),
  finishDate: PropTypes.instanceOf(dayjs),
  saved: PropTypes.bool,
};

export default TaskForm;
