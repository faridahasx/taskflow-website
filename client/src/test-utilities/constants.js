import { MISSING_INPUT_FIELD } from "constants/alertMessages";

// AUTH
export const passwordLabel = "Password:";
export const emailLabel = "Email:";
export const firstnameLabel = "First Name:";
export const lastnameLabel = "Last Name:";

// Server responses
export const successResponse = { status: 200, data: "Success" };

// Redux state
export const reduxStateOnMissingInputField = [
  {
    type: "ALERT",
    payload: MISSING_INPUT_FIELD,
  },
];
