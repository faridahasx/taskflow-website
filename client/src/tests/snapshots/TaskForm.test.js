import { render } from "@testing-library/react";
import TaskForm from "../../components/Tasks/TaskForm/TaskForm";
import { categoriesSample } from "../../constants/sampleData";

const mockFn = jest.fn();

test("TaskForm snapshot", () => {
  const { asFragment } = render(
    <TaskForm
      handleCloseEditor={mockFn}
      handleSubmit={mockFn}
      handleInputChange={mockFn}
      handleSetStartDate={mockFn}
      handleSetFinishDate={mockFn}
      handleSetDescription={mockFn}
      category=""
      categories={categoriesSample}
      loading={false}
    />
  );

  expect(asFragment()).toMatchSnapshot("TaskForm");
});
