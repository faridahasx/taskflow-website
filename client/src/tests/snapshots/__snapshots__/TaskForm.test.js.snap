// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm snapshot: TaskForm 1`] = `
<DocumentFragment>
  <div
    class="modal task-modal"
  >
    <div
      class="task-dialog"
    >
      <div
        class="flex task-dialog-top"
      >
        <button
          class="icon "
          title="Close"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CloseIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
        <button
          class="icon "
          form="task-form"
          title="Save"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DoneIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
            />
          </svg>
        </button>
      </div>
      <form
        class="column"
        id="task-form"
      >
        <input
          class="task-dialog-title"
          maxlength="200"
          name="title"
          placeholder="Title..."
          required=""
          type="text"
          value=""
        />
        <div
          class="description-input-loading center"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
            data-testid="loading"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
        </div>
        <div
          class="flex"
          id="select-task-details"
        >
          <div
            class="select-task-detail"
          >
            <div
              class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                id="category"
              >
                Category
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              >
                <div
                  aria-controls=":r0:"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="category category"
                  class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="category"
                  role="combobox"
                  tabindex="0"
                >
                  All
                </div>
                <input
                  aria-hidden="true"
                  aria-invalid="false"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  name="category"
                  tabindex="-1"
                  value="All"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      category
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="select-task-detail"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-7rtjrq-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r2:"
                id=":r2:-label"
              >
                Start Date
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl Mui-readOnly MuiInputBase-readOnly css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  aria-label="Choose date"
                  autocomplete="off"
                  class="MuiInputBase-input MuiOutlinedInput-input Mui-readOnly MuiInputBase-readOnly css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r2:"
                  inputmode="text"
                  placeholder="MM/DD/YYYY hh:mm aa"
                  readonly=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Start Date
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="select-task-detail"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-7rtjrq-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r5:"
                id=":r5:-label"
              >
                Finish Date
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl Mui-readOnly MuiInputBase-readOnly css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  aria-label="Choose date"
                  autocomplete="off"
                  class="MuiInputBase-input MuiOutlinedInput-input Mui-readOnly MuiInputBase-readOnly css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r5:"
                  inputmode="text"
                  placeholder="MM/DD/YYYY hh:mm aa"
                  readonly=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Finish Date
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
